man/hfst-pmatch2fst.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.49.3.
.TH HFST-PMATCH2FST "1" "April 2026" "HFST" "User Commands"
.SH NAME
hfst-pmatch2fst \- =Compile regular expressions into transducer(s)
.SH SYNOPSIS
.B hfst-pmatch2fst
[\fI\,OPTIONS\/\fR...] [\fI\,INFILE\/\fR]
.SH DESCRIPTION
Compile regular expressions into transducer(s)
.IP
(Experimental version)
.SS "Common options:"
.TP
\fB\-h\fR, \fB\-\-help\fR
Print help message
.TP
\fB\-V\fR, \fB\-\-version\fR
Print version info
.TP
\fB\-v\fR, \fB\-\-verbose\fR
Print verbosely while processing
.TP
\fB\-q\fR, \fB\-\-quiet\fR
Only print fatal erros and requested output
.TP
\fB\-s\fR, \fB\-\-silent\fR
Alias of \fB\-\-quiet\fR
.TP
\fB\-\-colour\fR[=\fI\,WHEN\/\fR]
Print in colour WHEN:
.TP
\fB\-\-color\fR[=\fI\,WHEN\/\fR]
always, never, auto (default)
.SS "Input/Output options:"
.TP
\fB\-i\fR, \fB\-\-input\fR=\fI\,INFILE\/\fR
Read input transducer from INFILE
.TP
\fB\-o\fR, \fB\-\-output\fR=\fI\,OUTFILE\/\fR
Write output transducer to OUTFILE
.SS "String and format options:"
.TP
\fB\-e\fR, \fB\-\-epsilon\fR=\fI\,EPS\/\fR
Map EPS as zero
.TP
\fB\-\-flatten\fR
Compile in all RTNs
.TP
\fB\-\-cosine\-distances\fR
When compiling Like() operations, include cosine distance info
.PP
If OUTFILE or INFILE is missing or \-, standard streams will be used.
If EPS is not defined, the default representation of 0 is used
Weights are currently not implemented.
.SH EXAMPLES
.TP
echo "Define TOP
UppercaseAlpha Alpha* LC({professor}) EndTag(ProfName);" | hfst\-pmatch2fst
.IP
create matcher that tags "professor Chomsky" as "professor <ProfName>Chomsky</ProfName>"
.SH "REPORTING BUGS"
Report bugs to <hfst\-bugs@helsinki.fi> or directly to our bug tracker at:
<https://github.com/hfst/hfst/issues>
.PP
hfst\-pmatch2fst home page:
<https://github.com/hfst/hfst/wiki/Pmatch2Fst>
.br
General help using HFST software:
<https://github.com/hfst/hfst/wiki>
.SH COPYRIGHT
Copyright \(co 2017 University of Helsinki,
License GPLv3: GNU GPL version 3 <http://gnu.org/licenses/gpl.html>
.br
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.


NEWS
# NEWS for hfst-tools and library

This file contains all noteworthy changes in HFST development between releases.
For full listing of changes see ChangeLog.

## Noteworthy changes in 3.17.1

* fixes logic error where `hfst-lexc` would treat a warning about implicit
  alphabet in XREs as a critical error even when `-Werror` is not asked

## Noteworthy changes in 3.17.0

* API breaks: HfstTokenizer, LexcCompiler, `read_in_att_format` functions
  in HfstTransducer, HfstBasicTransducer, etc. to accommodate the following
  unicode handling changes and error reporting:
* Character "tokenisers" that used to operate on bytes or UTF-8 sequences will
  now operate on [graphemes](https://www.unicode.org/reports/tr29/tr29-23.html#Grapheme_Cluster_Boundaries),
  this means that by default arcs in FSAs are expected to match combining
  grapheme clusters, e.g. a base character plus all diacritics.
  The implementation is based on ICU implementation of `u_brk`
    * in lexc string entries, graphemes are compiled into a arcs
    * in xerox regex curly braces, as well
    * in tokenise without --multichar mode, input is tokenised by graphemens
    * etc. etc. tbd
* negative weights yield warnings now unless turned off
* tools will have more gcc-style warning control: -Wall -Werror -Wno-this-error
* lexc warnings are systematically treated as errors with -Werror now
* colourised more warnings and errors

## Noteworthy changes in 3.16.2

* Support for `>OpenFst-1.8.0`
    * only functional change is removal of n-best paths search for log weight
      automata, it probably has errored out before but has slightly different
      error now (HfstException instead of openfst exit)
* Optimisations to hfst-pmatch2fst alphabet handling

## Noteworthy changes in 3.16.1

* Bug fixes:

    * hfst-tokenize escapes both \r and \n now

    * regexp parsing more robust with comments and linebreaks

* Removal of some bundled libraries including foma and openfst, install
  suitable versions to compile support for them

* ICU stabilised as unicode handler over all

* experimental C API for rust bindings (unstable)

* new tools (testing):

    * hfst-check-alpha for checking and comparing alphabets in automata and
      their metadata

    * hfst-dump-alphabets for reviewing alphabets easier than hfst-summarize
      grepping

    * hfst-kill-paths for efficient path removal

* introduced clang-format to work towards standard code formatting

Noteworthy changes in 3.16.0
----------------------------

* Bugfixes:

  * During conversion between optimized-lookup and other formats, the alphabet
    was not being preserved, leading to incorrect behaviour in some cases.
    This has been fixed.

  * Fix segfaults in hfst-summarize, hfst-disjunct and foma back-end.

  * Force ICU, fixes #501.

  * Use unique_ptr for input/ouput stream memory allocation (#547).


* Improvements:

  * hfst-pmatch archives now no longer include or parse redundant alphabets,
    which in some cases can be a meaningful size and speed saving.

  * Multiple memory leaks fixed.

  * Drop support for TR1 standard assuming that C++14 is minimal requirement.

  * Drop legacy wrappers for foma and lexc (#532).

  * Use upstream Foma and by default upstream OpenFST.

  * Use C++17, or C++14 with bundled OpenFST.

  * Update external links for sfst, foma and flex.

  * Add lex/yacc sources to dist (see issue #524).

* Features:

  * A new variable in the hfst-pmatch language, "unicode-character-classes",
    which causes the use of built-in character sets to be interpreted at
    runtime using a unicode library.

  * Verbose output from hfst-pmatch2fst has been much improved, providing eg.
    a call graph of the process. It also warns about use of pmatch features
    that is likely to be incorrect and lead to errors.


Noteworthy changes in 3.15.5
----------------------------

* Fixes to pmatch:

  * Modify tracking of entry points to allow for RTNs in Captures()

  * Fix off-by-one bug in checking captures

* Fixes to other issues: #508, #488, #486, #452, #491, #494

* Fix memory leaks


Noteworthy changes in 3.15.4
----------------------------

* Import Foma patches with license change to Apache-2.0 and minimal other fixes

* Allow regeneration of Foma files so the generated files can be omitted

* Explicit license file for SFST


Noteworthy changes in 3.15.3
----------------------------

* Numerous improvements to command line tools, e.g. substitute, insert-freely, proc, fst2txt, expand


Noteworthy changes in 3.15.2
----------------------------

* Fix off-by-one bug in checking captures in pmatch


Noteworthy changes in 3.15.1
----------------------------

* Fix issue #443

* Fix issue #429

* Fix issue #409


Noteworthy changes in 3.15.0
----------------------------

* Numerous improvements in pmatch processing:

  * hfst-pmatch2fst now implements eg @include"filename.txt" as a preprocessing stage
  
  * hfst-pmatch2fst now implements Explode() and Implode() in full generality
  
  * hfst-pmatch now allows printing weights (in location mode, with --locate)
  
  * hfst.PmatchContainer.locate offered via python bindings
  
  * bugfixes and improvements hfst.PmatchContainer.tokenize

* Add tool hfst-eliminate-flags.

* Allow both escaped and unescaped zeros in multicharacters in all lexc lexicon entries (fixes issue #274).

* Also reweight state that is both initial and final in hfst-reweight (fixes issue #368).

* Do not let distributed files depend on configuration. Include man files in dist.

* Print warnings and error messages also in earlier stages of twol compilation.

* Add option --execute-and-quit (-E) to hfst-xfst.


Noteworthy changes in 3.14.0
----------------------------

* Numerous improvements to pmatching and tokenization:

  * pmatch now supports the default symbol

  * pmatch now supports reading word embeddings in the binary format


tools/src/hfst-pmatch2fst.cc
//! @file hfst-pmatch2fst.cc
//!
//! @brief regular expression compiling command line tool
//!
//! @author HFST Team


//  This program is free software: you can redistribute it and/or modify
//  it under the terms of the GNU General Public License as published by
//  the Free Software Foundation, version 3 of the License.
//
//  This program is distributed in the hope that it will be useful,
//  but WITHOUT ANY WARRANTY; without even the implied warranty of
//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//  GNU General Public License for more details.
//
//  You should have received a copy of the GNU General Public License
//  along with this program.  If not, see <http://www.gnu.org/licenses/>.

#ifdef HAVE_CONFIG_H
#  include <config.h>
#endif

#ifdef WINDOWS
#include <io.h>
#endif

#ifndef WINDOWS
#include <unistd.h>
#endif

#include <iostream>
#include <fstream>
#include <memory>

#include <vector>
#include <map>
#include <string>
#include <iomanip>

using std::string;
using std::vector;
using std::pair;

#include <cstdio>
#include <cstdlib>
#include <cstring>

#ifdef _MSC_VER
#  include "hfst-getopt.h"
#else
#  include <getopt.h>
#endif

#include <math.h>
#include <errno.h>
#include <time.h>

#include "HfstTransducer.h"
#include "HfstInputStream.h"
#include "HfstOutputStream.h"
#include "HfstExceptionDefs.h"
#include "implementations/ConvertTransducerFormat.h"
#include "parsers/PmatchCompiler.h"
#include "hfst-commandline.h"
#include "hfst-program-options.h"
#include "hfst-tool-metadata.h"

#include "inc/globals-common.h"
#include "inc/globals-unary.h"

using hfst::HfstOutputStream;
using hfst::HfstTokenizer;
using hfst::HfstTransducer;
using hfst::pmatch::PmatchCompiler;

static char *epsilonname=NULL;
static bool disjunct_expressions=false;
static bool line_separated = false;
static bool flatten = false;
static bool include_cosine_distances = false;
static clock_t timer;

#if HAVE_OPENFST
static hfst::ImplementationType compilation_format = hfst::TROPICAL_OPENFST_TYPE;
#elif HAVE_FOMA
static hfst::ImplementationType compilation_format = hfst::FOMA_TYPE;
#elif HAVE_SFST
static hfst::ImplementationType compilation_format = hfst::SFST_TYPE;
#else
static hfst::ImplementationType compilation_format = hfst::ERROR_TYPE;
#endif

void
print_usage()
{
    // c.f. http://www.gnu.org/prep/standards/standards.html#g_t_002d_002dhelp
    fprintf(message_out, "Usage: %s [OPTIONS...] [INFILE]\n"
            "Compile regular expressions into transducer(s)\n (Experimental version)"
            "\n", program_name);
    print_common_program_options(message_out);
    print_common_unary_program_options(message_out);
    fprintf(message_out, "String and format options:\n"
            "  -e, --epsilon=EPS         Map EPS as zero\n"
            "      --flatten             Compile in all RTNs\n"
            "      --cosine-distances    When compiling Like() operations, include cosine distance info\n");
    fprintf(message_out, "\n");

    fprintf(message_out,
            "If OUTFILE or INFILE is missing or -, standard streams will be used.\n"
            "If EPS is not defined, the default representation of 0 is used\n"
            "Weights are currently not implemented.\n"
            "\n"
        );

    fprintf(message_out, "Examples:\n"
            "  echo \"Define TOP  UppercaseAlpha Alpha* LC({professor}) EndTag(ProfName);\" | %s \n"
            "  create matcher that tags \"professor Chomsky\" as \"professor <ProfName>Chomsky</ProfName>\"\n"
            "\n", program_name);
    print_report_bugs();
    fprintf(message_out, "\n");
    print_more_info();
    fprintf(message_out, "\n");
}

int
parse_options(int argc, char** argv)
{
    extend_options_getenv(&argc, &argv);
    // use of this function requires options are settable on global scope
    while (true)
    {
        static const struct option long_options[] =
            {
                HFST_GETOPT_COMMON_LONG,
                HFST_GETOPT_UNARY_LONG,
                {"epsilon", required_argument, 0, 'e'},
                {"flatten", no_argument, 0, '1'},
                {"cosine-distances", no_argument, 0, '2'},
                {0,0,0,0}
            };
        int option_index = 0;
        int c = getopt_long(argc, argv, HFST_GETOPT_COMMON_SHORT
                             HFST_GETOPT_UNARY_SHORT "e:",
                             long_options, &option_index);
        if (-1 == c)
        {
            break;
        }

        switch (c)
        {
#include "inc/getopt-cases-common.h"
#include "inc/getopt-cases-unary.h"
        case 'e':
            epsilonname = hfst_strdup(optarg);
            break;
        case '1':
            flatten = true;
            break;
        case '2':
            include_cosine_distances = true;
            break;
#include "inc/getopt-cases-error.h"
        }
    }

#include "inc/check-params-common.h"
#include "inc/check-params-unary.h"
    return EXIT_CONTINUE;
}

#ifndef _GNU_SOURCE
char * get_current_dir_name()
{
    size_t PATH_BUFSIZE = 1024;
    size_t PATH_BUFSIZE_MAX = PATH_BUFSIZE * 8; // sanity
    char * retval = (char *) NULL;
    while(retval == NULL) {
        retval = (char *) realloc(retval, PATH_BUFSIZE);
        if(getcwd(retval, PATH_BUFSIZE) != 0) {
            return retval;
        }
        if(errno == ERANGE && PATH_BUFSIZE < PATH_BUFSIZE_MAX) {
            PATH_BUFSIZE *= 2;
            continue;
        }
        if (errno == EACCES) {
            throw std::runtime_error("Unable to access working directory");
        }
        // something else went wrong, just forget about it and try to go on
        retval[0] = '\0';
        return retval;
    }
}
#endif

int
process_stream(HfstOutputStream& outstream)
{
    PmatchCompiler comp(compilation_format);
    comp.set_verbose(verbose);
    comp.set_flatten(flatten);
    comp.set_include_cosine_distances(include_cosine_distances);
    std::string file_contents;
    std::map<std::string, HfstTransducer*> definitions;
    int c;

#ifdef _MSC_VER
    if (inputfile == stdin && !silent)
      {
        warning(0, 0, "Reading from standard input. UTF-8 characters\n"
                "outside ascii range are supported only if input comes from a file.");
      }
#endif

    std::string includedir = "";
#ifndef _MSC_VER
    std::string inputfilename_str(inputfilename);
    if (inputfile != stdin && inputfilename_str.size() > 0) {
        if (inputfilename_str[0] == '/') {
            // absolute path
            includedir = inputfilename_str;
        } else {
            char * pwd = get_current_dir_name();
            std::string tmp(pwd);
            includedir = tmp + "/" + inputfilename_str;
            free(pwd);
        }
        size_t slashpos = includedir.rfind('/');
        if (slashpos == std::string::npos) {
            // mysterious, we'll just use the working dir
            includedir = "";
        } else {
            includedir = includedir.substr(0, slashpos + 1);
        }
    }
#endif
    comp.set_include_path(includedir);

    while ((c = fgetc(inputfile)) != EOF) {
        file_contents.push_back(c);
    }
    if (file_contents.size() > 1) {
      try
        {
          definitions = comp.compile(file_contents);
        }
      catch(HfstException & e)
        {
          std::cerr << e.name << std::endl;
          return EXIT_FAILURE;
        }
    }

    if (verbose) {
        std::cerr << std::setiosflags(std::ios::fixed) << std::setprecision(2);
        timer = clock();
        std::cerr << "Building hfst-ol alphabet... ";
    }
    
    // A dummy transducer with an alphabet with all the symbols
    HfstTransducer harmonizer(compilation_format);
    
    // First we need to collect a unified alphabet from all the transducers.
    hfst::StringSet symbols_seen;
    for (std::map<std::string, HfstTransducer *>::const_iterator it =
             definitions.begin(); it != definitions.end(); ++it) {
        hfst::StringSet string_set = it->second->get_alphabet();
        for (const auto & sym : string_set) {
            if (symbols_seen.count(sym) == 0) {
                harmonizer.disjunct(HfstTransducer(sym, compilation_format));
                symbols_seen.insert(sym);
            }
        }
    }
    if (symbols_seen.size() == 0) {
        // We don't recognise anything, go home early
        std::cerr << program_name << ": Empty ruleset, nothing to write\n";
        return EXIT_FAILURE;
    }
    
    // Then we convert it...
    harmonizer.convert(hfst::HFST_OLW_TYPE);
    // Use these for naughty intermediate steps to make sure
    // everything has the same alphabet
    hfst::HfstBasicTransducer * intermediate_tmp;
    hfst_ol::Transducer * harmonized_tmp;
    hfst::HfstTransducer * output_tmp;
    
    if (verbose) {
        double duration = (clock() - timer) /
            (double) CLOCKS_PER_SEC;
        timer = clock();
        std::cerr << "built in " << duration << " seconds\n";
        std::cerr << "Converting TOP... ";
    }
        
    // When done compiling everything, look for TOP and output it first.
    if (definitions.count("TOP") == 1) {
        std::map<std::string, std::string> properties = definitions["TOP"]->get_properties();
        intermediate_tmp = hfst::implementations::ConversionFunctions::
            hfst_transducer_to_hfst_basic_transducer(*definitions["TOP"]);
        harmonized_tmp = hfst::implementations::ConversionFunctions::
            hfst_basic_transducer_to_hfst_ol(intermediate_tmp,
                                             true, // weighted
                                             "", // no special options
                                             &harmonizer); // harmonize with this
        output_tmp = hfst::implementations::ConversionFunctions::
            hfst_ol_to_hfst_transducer(harmonized_tmp);
        output_tmp->set_name("TOP");
        for(auto & propertie : properties) {
            output_tmp->set_property(propertie.first, propertie.second);
        }
        outstream << *output_tmp;
        delete definitions["TOP"];
        definitions.erase("TOP");
        delete intermediate_tmp;
        delete output_tmp;

        if (verbose) {
            double duration = (clock() - timer) /
                (double) CLOCKS_PER_SEC;
            timer = clock();
            std::cerr << "converted in " << duration << " seconds\n";
        }
    
        for (auto & definition : definitions) {
            if (verbose) {
                std::cerr << "Converting " << definition.first << "... ";
                timer = clock();
            }
            intermediate_tmp = hfst::implementations::ConversionFunctions::
                hfst_transducer_to_hfst_basic_transducer(*(definition.second));
            harmonized_tmp = hfst::implementations::ConversionFunctions::
                hfst_basic_transducer_to_hfst_ol(intermediate_tmp,
                                                 true, // weighted
                                                 "", // no special options
                                                 &harmonizer); // harmonize with this
            output_tmp = hfst::implementations::ConversionFunctions::
                hfst_ol_to_hfst_transducer(harmonized_tmp);
            output_tmp->set_name(definition.first);
            outstream << *output_tmp;;
            delete definition.second;
            delete intermediate_tmp;
            delete output_tmp;
            if (verbose) {
                double duration = (clock() - timer) /
                    (double) CLOCKS_PER_SEC;
                std::cerr << "converted in " << duration << " seconds\n";
            }
        }
    } else {
        std::cerr << program_name << ": Empty ruleset, nothing to write\n";
        return EXIT_FAILURE;
    }
    outstream.close();
    return EXIT_SUCCESS;
}

extern int pmatchdebug;

int main( int argc, char **argv )
{
#ifdef WINDOWS
  _setmode(1, _O_BINARY);
#endif

//    pmatchdebug = 1;

    hfst_set_program_name(argv[0], "0.1", "Pmatch2Fst");
    int retval = parse_options(argc, argv);
    if (retval != EXIT_CONTINUE)
    {
        return retval;
    }
    // close buffers, we use streams
    if (outfile != stdout)
    {
        fclose(outfile);
    }
    verbose_printf("Reading from %s, writing to %s\n",
                   inputfilename, outfilename);
    // here starts the buffer handling part
    auto outstream = (outfile != stdout) ?
        std::make_unique<HfstOutputStream>(outfilename, hfst::HFST_OLW_TYPE) :
        std::make_unique<HfstOutputStream>(hfst::HFST_OLW_TYPE);
    process_stream(*outstream);
    free(inputfilename);
    free(outfilename);
    return EXIT_SUCCESS;
}

